src/games/strategy/triplea/delegate/MoveValidator.java

```diff
@@ -1409,47 +1422,58 @@ public static String validateCanal(final Route route, final Collection<Unit> uni
    */
   public static String validateCanal(final Territory territory, final Route route, final Collection<Unit> units,
       final PlayerID player, final GameData data) {
+    String failureMessage = null;
     final Set<CanalAttachment> canalAttachments = CanalAttachment.get(territory);
-    if (canalAttachments.isEmpty()) {
-      return null;
-    }
-    final Iterator<CanalAttachment> iter = canalAttachments.iterator();
-    while (iter.hasNext()) {
-      final CanalAttachment attachment = iter.next();
-      if (attachment == null) {
+    for (final CanalAttachment canalAttachment : canalAttachments) {
+      if (!needToCheckCanal(canalAttachment, route, data)) {
```

Comment:
Having multiple commits for the refactor would have made the review easier. I take some objection since I feel a bit of a double standard, harsher comments in my PRs about things not belonging, while here we have everything in one commit..